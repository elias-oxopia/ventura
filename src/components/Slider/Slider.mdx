---
name: Slider
menu: Components
route: /slider
---

import { Playground, Props } from 'docz';
import { Slider } from '../../index.ts'

# Slider

A slider component lets the user pick a value between a minimum and maximum value.

## Examples

### Minimal slider

<Playground>
  {() => {
    const [value, setValue] = React.useState(50);
    return <Slider name='minimal-slider' valueBoundaries={[0,100]} value={value} onChange={setValue} />
  }}
</Playground>

### Slider with decimal values

<Playground>
  {() => {
    const [value, setValue] = React.useState(1);
    return <Slider name='minimal-slider' valueBoundaries={[0,10]} value={value} onChange={setValue} numberOfDecimals={2} stepSize={0.05} />
  }}
</Playground>

### Slider with text boundaries

<Playground>
  {() => {
    const [value, setValue] = React.useState(5);
    return <Slider name='minimal-slider' valueBoundaries={[0,10]} textBoundaries={['Not accurate at all', 'Very accurate']} value={value} stepSize={1} onChange={setValue} />
  }}
</Playground>

### Slider without tooltip

<Playground>
  {() => {
    const [value, setValue] = React.useState(50);
    return <Slider name='minimal-slider' valueBoundaries={[0,100]} value={value} onChange={setValue} showTooltip={false} />
  }}
</Playground>

### Disabled slider

<Playground>
  {() => {
    const [value, setValue] = React.useState(50);
    return <Slider name='minimal-slider' valueBoundaries={[0,100]} value={value} onChange={setValue} isDisabled />
  }}
</Playground>

## API

<Props of={Slider} />
