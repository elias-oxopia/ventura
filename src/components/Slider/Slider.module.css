.sliderWrapper {
  box-sizing: content-box;
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 20px;
  touch-action: none;
}

.slider {
  position: relative;
  width: 100%;
  height: 10px;
  margin-top: 5px;
}

.sliderWrapper.disabled {
  background-color: var(--colorBaseNeutral300);
}
