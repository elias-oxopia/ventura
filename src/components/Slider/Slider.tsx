import React from 'react';
import { Handles, Rail, Slider as CompoundSlider, Ticks } from 'react-compound-slider';
import classnames from 'classnames';

import { Handle } from './Handle/Handle';
import { SliderRail } from './SliderRail/SliderRail';
import { Tick } from './Tick/Tick';

import cssReset from '../../css-reset.module.css';
import styles from './Slider.module.css';

type Props = {
  value: number;
  onChange: (value: number) => void;
  valueBoundaries: [number, number];
  textBoundaries?: [string, string];
  stepSize?: number;
  numberOfDecimals?: number;
  isDisabled?: boolean;
  showTooltip?: boolean;
  className?: string;
};

const Slider: React.FC<Props> = ({
  value,
  onChange,
  valueBoundaries,
  textBoundaries,
  stepSize,
  numberOfDecimals = 0,
  isDisabled = false,
  showTooltip = true,
  className,
}: Props) => {
  const formatValue = (val: number) =>
    (Math.round(val * (10 ** numberOfDecimals || 1)) / (10 ** numberOfDecimals || 1)).toFixed(
      numberOfDecimals,
    );
  const minValue = valueBoundaries[0];
  const maxValue = valueBoundaries[1];
  const classNames = classnames(
    cssReset.ventura,
    styles.sliderWrapper,
    {
      [styles.disabled]: isDisabled,
    },
    className,
  );

  return (
    <div className={classNames}>
      <CompoundSlider
        step={stepSize ?? (maxValue - minValue) / 100}
        domain={valueBoundaries}
        className={styles.slider}
        onChange={(values) => onChange(values[0])}
        values={[value]}
        disabled={isDisabled}
      >
        <Rail>{({ getRailProps }) => <SliderRail getRailProps={getRailProps} />}</Rail>
        <Handles>
          {({ handles, activeHandleID, getHandleProps }) => (
            <div className="slider-handles">
              {handles.map((handle) => (
                <Handle
                  key={handle.id}
                  handle={handle}
                  domain={valueBoundaries}
                  isActive={handle.id === activeHandleID}
                  formatValue={formatValue}
                  getHandleProps={getHandleProps}
                  isDisabled={isDisabled}
                  showTooltip={showTooltip}
                />
              ))}
            </div>
          )}
        </Handles>
        <Ticks count={2} values={valueBoundaries}>
          {({ ticks }) => (
            <div className="slider-ticks">
              {ticks.map((tick, index) => (
                <Tick
                  key={tick.id}
                  tick={tick}
                  formatValue={formatValue}
                  textValue={textBoundaries?.[index]}
                />
              ))}
            </div>
          )}
        </Ticks>
      </CompoundSlider>
    </div>
  );
};

export default Slider;
