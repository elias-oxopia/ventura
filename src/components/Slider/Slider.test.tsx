import * as React from 'react';
import { render } from '@testing-library/react';

import Slider from './Slider';

describe('Slider', () => {
  const defaultProps: React.ComponentProps<typeof Slider> = {
    onChange: jest.fn(),
    value: 50,
    valueBoundaries: [0, 100],
  };

  test('default snapshot', () => {
    const { asFragment } = render(<Slider {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('3 decimals should be displayed', () => {
    const { queryByText } = render(<Slider {...defaultProps} numberOfDecimals={3} />);
    expect(queryByText('100.000')).not.toBe(null);
  });

  test('text boundaries should be displayed', () => {
    const { queryByText } = render(<Slider {...defaultProps} textBoundaries={['left', 'right']} />);
    expect(queryByText('left')).not.toBe(null);
    expect(queryByText('right')).not.toBe(null);
  });
});
