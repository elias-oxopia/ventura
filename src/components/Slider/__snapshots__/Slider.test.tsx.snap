// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider default snapshot 1`] = `[Function]`;
